File: core/src/commonMain/kotlin/com/jght/sjrqromx/business/shipping/kmp_shipping_platform/features/quoting/domain/usecase/CalculateQuoteUseCase.kt
package com.jght.sjrqromx.business.shipping.kmp_shipping_platform.features.quoting.domain.usecase

import com.jght.sjrqromx.business.shipping.kmp_shipping_platform.features.quoting.domain.model.*
import com.jght.sjrqromx.business.shipping.kmp_shipping_platform.features.quoting.domain.repository.TariffRemoteService

public class CalculateQuoteUseCase(
    private val tariffService: TariffRemoteService
) {
    suspend operator fun invoke(request: QuoteRequest): QuoteResult {
        // Step 1: Rich Domain Validation
        request.validate()?.let { error ->
            return QuoteResult.Error(error)
        }

        // Step 2: Infrastructure / External Data fetching
        val remoteResult = tariffService.getRemoteMultiplier(request.destinationZipCode)
        
        if (remoteResult.isFailure) {
            return QuoteResult.Error(
                QuoteError(
                    type = QuoteErrorType.REMOTE_SERVICE_ERROR,
                    code = "TARIFAS_SERVICE_UNAVAILABLE",
                    message = "Could not get zone multiplier, please try again"
                )
            )
        }

        val remoteMultiplier = remoteResult.getOrThrow()

        // Step 3: Pure Domain Calculation via ShippingEngine
        val response = ShippingEngine.calculate(request, remoteMultiplier)

        return QuoteResult.Success(response)
    }
}

File: flutter_app/lib/features/quoting/domain/usecase/calculate_quote_use_case.dart
import '../model/quote_models.dart';
import '../model/shipping_engine.dart';
import '../repository/tariff_repository.dart';

class CalculateQuoteUseCase {
  final TariffRepository _repository;

  CalculateQuoteUseCase(this._repository);

  Future<QuoteResult> call(QuoteRequest request) async {
    // Step 1: Validation (Rich Domain logic could be added to QuoteRequest too)
    if (request.weightKg <= 0) {
      return QuoteError(
        type: QuoteErrorType.validationError,
        code: 'INVALID_WEIGHT',
        message: 'El peso debe ser mayor a 0',
      );
    }

    if (request.distanceKm <= 0) {
      return QuoteError(
        type: QuoteErrorType.validationError,
        code: 'INVALID_DISTANCE',
        message: 'La distancia debe ser mayor a 0',
      );
    }

    if (request.destinationZipCode.length != 5 || int.tryParse(request.destinationZipCode) == null) {
      return QuoteError(
        type: QuoteErrorType.validationError,
        code: 'INVALID_ZIP',
        message: 'El código postal debe tener 5 dígitos',
      );
    }

    try {
      // Step 2: Infrastructure / External Data
      final remoteMultiplier = await _repository.getRemoteMultiplier(request.destinationZipCode);

      // Step 3: Pure Domain Calculation via ShippingEngine
      final response = ShippingEngine.calculate(request, remoteMultiplier);

      return QuoteSuccess(response);
    } catch (e) {
      return QuoteError(
        type: QuoteErrorType.remoteServiceError,
        code: 'TARIFAS_SERVICE_UNAVAILABLE',
        message: 'No se pudo obtener el multiplicador de zona, intenta de nuevo',
      );
    }
  }
}
